services/director-v2/tests/unit/test_modules_dynamic_sidecar_docker_service_specs_volume_remover.py

```diff
@@ -1,10 +1,205 @@
+    return tmp_path / "unused_volume"
+
+
+@pytest.fixture
+async def unused_volume(
+    swarm_stack_name: str,
+    study_id: str,
+    node_uuid: str,
+    run_id: str,
+    unused_volume_path: Path,
+) -> AsyncIterator[DockerVolume]:
+    async with aiodocker.Docker() as docker_client:
+        source = _get_source(run_id, node_uuid, unused_volume_path)
+        volume = await docker_client.volumes.create(
+            {
+                "Name": source,
+                "Labels": {
+                    "node_uuid": node_uuid,
+                    "run_id": run_id,
+                    "source": source,
+                    "study_id": study_id,
+                    "swarm_stack_name": swarm_stack_name,
+                    "user_id": "1",
+                },
+            }
+        )
+
+        # attach to volume and create some files!!!
+
+        yield volume
+
+        try:
+            await volume.delete()
+        except aiodocker.DockerError:
+            pass
+
+
+@pytest.fixture
+async def used_volume(
+    swarm_stack_name: str,
+    study_id: str,
+    node_uuid: str,
+    run_id: str,
+    used_volume_path: Path,
+) -> AsyncIterator[DockerVolume]:
+    async with aiodocker.Docker() as docker_client:
+        source = _get_source(run_id, node_uuid, used_volume_path)
+        volume = await docker_client.volumes.create(
+            {
+                "Name": source,
+                "Labels": {
+                    "node_uuid": node_uuid,
+                    "run_id": run_id,
+                    "source": source,
+                    "study_id": study_id,
+                    "swarm_stack_name": swarm_stack_name,
+                    "user_id": "1",
+                },
+            }
+        )
+
+        container = await docker_client.containers.run(
+            config={
+                "Cmd": ["/bin/ash", "-c", "sleep 10000"],
+                "Image": "alpine:latest",
+                "HostConfig": {"Binds": [f"{volume.name}:{used_volume_path}"]},
+            },
+            name=f"using_volume_{volume.name}",
+        )
+        await container.start()
+
+        yield volume
+
+        await container.delete(force=True)
+        await volume.delete()
+
+
+@pytest.fixture
+async def used_volume_name(used_volume: DockerVolume) -> str:
+    volume = await used_volume.show()
+    return volume["Name"]
+
+
+@pytest.fixture
+async def unused_volume_name(unused_volume: DockerVolume) -> str:
+    volume = await unused_volume.show()
+    return volume["Name"]
+
+
+@pytest.fixture
+def missing_volume_name(run_id: str, node_uuid: str) -> str:
+    return _get_source(run_id, node_uuid, Path("/MISSING/PATH"))
+
+
+# TESTS
+
+
+async def test_sh_script_error_if_volume_is_used(
+    used_volume_name: str, docker_version: DockerVersion
+):
+    result = await run_command(docker_version, volume_names=[used_volume_name])
+    print(result)
+    assert "ERROR: Please check above logs, there was/were 1 error/s." in result
```

Comment:
CAREFUL with test checking messages!! We have discussed that many times!

TIPS:
 - check only key words or
 - mock the message and replace with a TEST_MESSAGE

Reply:
Yes, that is pretty clear to me.

The idea here is to check if the sh script breaks. These messages are controlled by us and some are generated by docker. As SAN pointed out, these might change if docker changes. But these should mitigate the risk.